File: scripts/build_homepage.py
#!/usr/bin/env python3
"""
Build Homepage
Scans dist/ for book folders and generates index.html with links.
"""

import re
from pathlib import Path

DIST_DIR = Path('./dist')
TEMPLATE_DIR = Path('./templates')
OUTPUT_FILE = DIST_DIR / 'index.html'


def get_book_folders():
    """Find all book folders (directories with index.html)."""
    books = []
    for item in DIST_DIR.iterdir():
        if item.is_dir() and (item / 'index.html').exists():
            books.append(item.name)
    return sorted(books)


def extract_title(slug):
    """Extract title from book's index.html <title> tag."""
    index_path = DIST_DIR / slug / 'index.html'
    try:
        html = index_path.read_text(encoding='utf-8')
        match = re.search(r'<title>([^<]+)</title>', html)
        if match:
            title = match.group(1)
            # Remove common suffixes (match JS behavior)
            title = title.replace(' — Analysis', '')
            return title
    except Exception:
        pass
    # Fallback: use slug as-is
    return slug


def generate_homepage():
    """Generate the homepage HTML."""
    books = get_book_folders()
    
    # Generate book links (match JS output format)
    links = []
    for slug in books:
        title = extract_title(slug)
        links.append(f'<li><a href="/{slug}/">{title}</a></li>')
    
    links_html = '\n'.join(links)
    
    # Read template
    template_path = TEMPLATE_DIR / 'homepage.html'
    template = template_path.read_text(encoding='utf-8')
    html = template.replace('{{BOOK_LINKS}}', links_html)
    
    OUTPUT_FILE.write_text(html, encoding='utf-8')
    print(f'Homepage generated with {len(books)} books')


if __name__ == '__main__':
    generate_homepage()

File: scripts/build-homepage.js
// Scan dist/ for book folders
// Each folder with index.html = one book
// Generate homepage with links to each book

const fs = require('fs');
const path = require('path');

const DIST = './dist';
const TEMPLATE = './templates/homepage.html';
const OUTPUT = './dist/index.html';

function getBookFolders() {
  return fs.readdirSync(DIST)
    .filter(f => {
      const stat = fs.statSync(path.join(DIST, f));
      return stat.isDirectory() && fs.existsSync(path.join(DIST, f, 'index.html'));
    });
}

function buildHomepage() {
  const books = getBookFolders();
  const template = fs.readFileSync(TEMPLATE, 'utf8');
  
  const links = books.map(slug => {
    // Extract title from the book's index.html <title> tag
    const html = fs.readFileSync(path.join(DIST, slug, 'index.html'), 'utf8');
    const titleMatch = html.match(/<title>([^<]+)<\/title>/);
    const title = titleMatch ? titleMatch[1].replace(' — Analysis', '') : slug;
    
    return `<li><a href="/${slug}/">${title}</a></li>`;
  }).join('\n');
  
  const output = template.replace('{{BOOK_LINKS}}', links);
  fs.writeFileSync(OUTPUT, output);
  console.log(`Homepage generated with ${books.length} books`);
}

buildHomepage();

